hibernate-core/src/main/java/org/hibernate/query/criteria/internal/compile/CriteriaQueryTypeQueryAdapter.java

```diff
@@ -443,21 +443,13 @@ private ExplicitParameterInfo locateParameterByPosition(int position) {
 
 	public Parameter<?> getParameter(String name) {
 		entityManager.checkOpen( false );
-		return locateParameterByName( name );
+		return jpqlQuery.getParameter( name );
 	}
 
 	@Override
-	@SuppressWarnings({ "unchecked" })
 	public <T> Parameter<T> getParameter(String name, Class<T> type) {
 		entityManager.checkOpen( false );
-		Parameter parameter = locateParameterByName( name );
-		if ( type.isAssignableFrom( parameter.getParameterType() ) ) {
-			return parameter;
-		}
-		throw new IllegalArgumentException(
-				"Named parameter [" + name + "] type is not assignanle to request type ["
-						+ type.getName() + "]"
-		);
```

Comment:
I think the type assignable check is necessary, base on the javax.persistence.Query javadoc ```
/**
     * Get the parameter object corresponding to the declared
     * parameter of the given name and type.
     * This method is required to be supported for criteria queries
     * only.
     * @param name  parameter name
     * @param type  type
     * @return parameter object
     * @throws IllegalArgumentException if the parameter of the
     *         specified name does not exist or is not assignable
     *         to the type
     * @throws IllegalStateException if invoked on a native
     *         query or Java Persistence query language query when
     *         the implementation does not support this use
     * @since Java Persistence 2.0
     */
    <T> Parameter<T> getParameter(String name, Class<T> type);
```
 an IllegalArgumentException has to be thrown if the parameter of the specified name does not exist or is not assignable to the type.
The  removed `locateParameterByName( name )` was throwing the exception in case the specified name does not exists.

Reply:
Yes, I agree with you. It is still making this check, the only difference is that, it is delegated to the underlying jpqlQuery object `return jpqlQuery.getParameter( name, type );`